### PR TITLE
TOBAGO-2021: Sheet should be able to lazy load rows by scroll events

### PR DESCRIPTION
**Pull request by lofwyr14**

### CODE REVIEW

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 1]
```diff
@@ -426,7 +494,69 @@ Tobago.Sheet.setup2 = function (sheets) {
...
+      tr.setAttribute("dummy", "dummy");
+      tr.style.height = height + "px";
+      var td = document.createElement("td");
+      td.classList.add("tobago-sheet-cell");
+      td.colSpan = columns;
+      tr.appendChild(td);
+      return tr;
+    }
+
+    console.time("lazy");
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: dev_only missing

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 2]
```diff
@@ -90,24 +93,89 @@ Tobago.Sheet.prototype.reloadWithAction = function(source, action) {
...
+
+        var sheet = document.getElementById(id);
+        sheet.id = id + "::lazy-temporary";
+
+        var page = Tobago.findPage();
+        page.get(0).insertAdjacentHTML("beforeend", "<div id='" + id + "'></div>");
+        var sheetLoader = document.getElementById(id);
+      }
+    }
+  } else   if (event.status === "success") {
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: formatting code

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 3]
```diff
@@ -426,7 +494,69 @@ Tobago.Sheet.setup2 = function (sheets) {
...
+    if (lazy) {
+      // prepare the sheet with some auto-created (empty) rows
+      var rows = $sheet.data("tobago-rows");
+      var rowCount = $sheet.data("tobago-row-count");
+      var $sheetBody = $sheet.find(".tobago-sheet-body");
+      var $tbody = $sheetBody.find("tbody");
+      var columns = $tbody.find("tr:first").find("td").length;
+      var height = $tbody.height() / rows;
+      var pointer = $tbody.get(0).rows[0]; // points current row
+      console.time("lazy2");
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: dev_only missing

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 4]
```diff
@@ -426,7 +494,69 @@ Tobago.Sheet.setup2 = function (sheets) {
...
+            pointer.insertAdjacentElement("afterend", template);
+            pointer = template;
+          }
+        } else {
+          template = getTemplate(height, columns, i);
+          $tbody.get(0).insertAdjacentElement("beforeend", template);
+          pointer = template;
+        }
+      }
+      console.timeEnd("lazy2");
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: dev_only missing

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 5]
```diff
@@ -426,7 +494,69 @@ Tobago.Sheet.setup2 = function (sheets) {
...
+      $sheetBody.bind("scroll", function () {
+        Tobago.Sheet.lazyCheck($sheet);
+      });
+
+      // initial
+      Tobago.Sheet.lazyCheck($sheet);
+
+      $sheet.data("tobago-lazy-initialized", true);
+    }
+    console.timeEnd("lazy");
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: dev_only missing

File: tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js [hunk 6]
```diff
@@ -638,12 +840,7 @@ Tobago.Sheet.selectRange = function($sheet, $rows, first, last, selectDeselected
 };
 
 Tobago.Sheet.getDataIndex = function(sheet, row) {
-  var rowIndex = row.data("tobago-row-index");
-  if (rowIndex) {
-    return +rowIndex;
-  } else {
-    return row.index() + sheet.data("tobago-first");
-  }
+  return row.data("tobago-row-index")
```
Comment (henningn) [inline tobago-theme/tobago-theme-standard/src/main/resources/META-INF/resources/tobago/standard/tobago-bootstrap/_version/js/tobago-sheet.js]: missing ';'

**Comment by lofwyr14** [comment]
fixed suggestions from @henningn

**Pull request merged**